From 56382f8d9a80796a21d7087ddb660519fbbb5c2e Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Wed, 16 May 2018 10:21:57 +0100
Subject: Process DBus replies in the ::match thread
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Summary:
6114612970bcf757d9e06a623d6220257ce57b5f turned the DBus requests into
using QDBusPendingCallWatcher but blocking so that it's still handled
within the ::match() method. However, it accidentally processed replies
in the main thread, rather than in the match thread which wasn't the
intention.

Result is occasional threading related crashes.

BUG: 394272

Test Plan:
Still had results from baloo (one of our DBus runners)
It didn't crash (but it never crashed for me before either)
‎
One tester replied:
[10:00] ‎<‎sunnyflunk‎>‎ d_ed: looks good, tried 30 times with patch and no crash, then turned off the patch and crashed 4th time

Reviewers: #plasma, broulik

Reviewed By: #plasma, broulik

Subscribers: kde-frameworks-devel

Tags: #frameworks

Differential Revision: https://phabricator.kde.org/D12924
---
 src/dbusrunner.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/dbusrunner.cpp b/src/dbusrunner.cpp
index f89b05e..65ae486 100644
--- a/src/dbusrunner.cpp
+++ b/src/dbusrunner.cpp
@@ -160,7 +160,7 @@ void DBusRunner::match(Plasma::RunnerContext &context)
 
                 context.addMatch(m);
             };
-        });
+        }, Qt::DirectConnection); // process reply in the watcher's thread (aka the one running ::match  not the one owning the runner)
     }
     //we're done matching when every service replies
     for (auto w : watchers) {
--
cgit v0.11.2
